pkg/components/linkerd/component.go

```diff
@@ -0,0 +1,127 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package linkerd
+
+// nolint: gci
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/gohcl"
+	"github.com/linkerd/linkerd2/pkg/tls"
+
+	internaltemplate "github.com/kinvolk/lokomotive/internal/template"
+	"github.com/kinvolk/lokomotive/pkg/components"
+	"github.com/kinvolk/lokomotive/pkg/components/util"
+	"github.com/kinvolk/lokomotive/pkg/k8sutil"
+)
+
+const (
+	name = "experimental-linkerd"
+)
+
+// nolint:gochecknoinits
+func init() {
+	components.Register(name, newComponent())
+}
+
+type component struct {
+	ControllerReplicas int `hcl:"controller_replicas,optional"`
+
+	CA     string
+	Cert   string
+	Key    string
+	Expiry string
+}
+
+func newComponent() *component {
+	return &component{
+		ControllerReplicas: 1,
+	}
+}
+
+func (c *component) LoadConfig(configBody *hcl.Body, evalContext *hcl.EvalContext) hcl.Diagnostics {
+	diagnostics := hcl.Diagnostics{}
+
+	if configBody == nil {
+		return hcl.Diagnostics{}
+	}
+
+	d := gohcl.DecodeBody(*configBody, evalContext, c)
+	if d.HasErrors() {
+		return append(diagnostics, d...)
+	}
+
+	return diagnostics
+}
+
+func (c *component) RenderManifests() (map[string]string, error) {
+	helmChart, err := util.LoadChartFromAssets("/components/linkerd2")
+	if err != nil {
+		return nil, fmt.Errorf("load chart from assets: %w", err)
+	}
+
+	// Generate certs.
+	root, err := tls.GenerateRootCAWithDefaults("identity.linkerd.cluster.local")
+	if err != nil {
+		return nil, fmt.Errorf("could not generate cert: %w", err)
+	}
+
+	c.Key = indent(root.Cred.EncodePrivateKeyPEM(), 8)
+	c.Cert = indent(root.Cred.Crt.EncodeCertificatePEM(), 8)
+	c.CA = indent(root.Cred.Crt.EncodeCertificatePEM(), 4)
+	c.Expiry = root.Cred.Crt.Certificate.NotAfter.String()
+
+	values, err := internaltemplate.Render(chartValuesTmpl, c)
+	if err != nil {
+		return nil, fmt.Errorf("rendering values template failed: %w", err)
+	}
+
+	// Generate YAML for the Linkerd deployment.
+	renderedFiles, err := util.RenderChart(helmChart, name, c.Metadata().Namespace.Name, values)
+	if err != nil {
+		return nil, fmt.Errorf("rendering chart failed: %w", err)
+	}
+
+	return renderedFiles, nil
+}
+
+func (c *component) Metadata() components.Metadata {
+	return components.Metadata{
+		Name: name,
+		Namespace: k8sutil.Namespace{
+			Name: "linkerd",
+		},
+		Helm: components.HelmMetadata{
+			Wait: true,
+		},
+	}
+}
+
+func indent(data string, indent int) string {
```

Comment:
Perhaps we should put it in `internal` package, so it can be potentially re-used by other components.

Reply:
Also, for following inputs, this function does not work as expected:
- For `foo\nbar\n`, should return "  foo\n  bar\n", but returns "  foo\n  bar\n  " (notice trailing indent)
- For ``, should IMO return ``, but returns `  `.

I think we should add some tests for it.